### ISSUE TITLE
ADD delay to sync with audio

### ISSUE THREAD

**Issue by ZornTaov** [emgucv/emgucv#547]
### Discussed in https://github.com/emgucv/emgucv/discussions/546

<div type='discussions-op-text'>

<sup>Originally posted by **ZornTaov** July 21, 2021</sup>
Preface:
I'm working on a project that captures the Video and Audio inputs from (hopefully) Any game capture device and render it on a window with customizable inputs and outputs for video in and audio in/out.  What this does is it allows me to stream whatever console game I'm playing over Discord and have the viewers be able to see and hear the game (as opposed to using OBS or the like to do so, as discord has trouble capturing both audio and video with those, and I don't want to mix the game audio with my microphone audio directly).

The Problem I need help with:
On most capture devices, there will always be a bit of latency between what happens in the game and what gets rendered to the programs window.  With some there is also added latency with either the audio or video, causing either to be out of sync with each other.  I am using NAudio for the audio streaming and I am able to add a delay to it if needed, but I can't figure out a way to delay the video capture if the audio lags behind (example, I fire a gun in game, the video shows me firing, seconds later the gunshot is heard).

Code Source:
https://github.com/ZornTaov/GameCaptureForDiscord/
Render Code: https://github.com/ZornTaov/GameCaptureForDiscord/blob/master/VideoCaptureForm.cs#L115
Video Device Setup Code: https://github.com/ZornTaov/GameCaptureForDiscord/blob/master/ControlsForm.cs#L193-L220

Libraries used:
![image](https://user-images.githubusercontent.com/2144351/126452956-9bdd85c1-0506-4f9c-9c56-ac543fc464f7.png)
![image](https://user-images.githubusercontent.com/2144351/126452992-3cfa6f55-7bf0-4d2a-a70e-689538a1eade.png)
(unsure if I need some of those libraries or not)

Other Lesser Problems(will make an Issue if needed): 
Currently, I can easily render 1920x1080@30fps, which is just fine for my uses.  I might add to the ControlsForm a resolution selection combobox, but I need a way of getting the list of possible resolutions from a device.

Creating a new VideoCapture object each time I change a video device causes the application to hang for 30 seconds, and I can see LOTS of threads opening and closing in the console.  I think it has to do with using VideoCapture.API.Msmf?

Tl;Dr:
I need a way to add delay to video capture to sync with delayed audio. (and wasn't sure if this should be an issue or discussion)</div>

**Comment by emgucv** [comment]
You can implement it with a queue. Run capture on a separate thread, keep grabbing images and push the image into the queue with timestamp. 

When ever you need to query the delayed frame, check the front of the queue and see if it is from the "delay" amount of seconds ago, if it is, return it.

You need to keep monitoring the size of the queue and make sure it doesn't run out of memory.

**Comment by ZornTaov** [comment]
So I've implemented a sort of delayed queue using an Observable subscription from System.Reactive.  It WORKS, but it also drastically increases the amount of memory being used if a delay is added. With a 1080p, 30fps stream delayed by 10 seconds increases the programs memory footprint to around 2.1gb.  This is absolutely understandable, as that's around 18000 frames worth of data.  What I'd like to know now is 1) What kind of compression is being used by EMGU or OpenCV, AKA if either library is compressing the data or if it's the pre-compressed data from the video capture device, and 2) If there is a to further compress the video image data for storage before it's used after the delay. https://github.com/ZornTaov/GameCaptureForDiscord/blob/fb77e8c6f774f37b3459e907ff137c270a355c44/VideoCaptureForm.cs#L143-L166

**Comment by emgucv** [comment]
To answer the follow-up questions:

**1. Compression in Emgu CV / OpenCV**

None. A `Mat` received from `VideoCapture` holds raw, uncompressed BGR pixel data — every pixel is 3 bytes. At 1920×1080 that is ~6 MB per frame, so a 10-second delay at 30 fps means ~300 frames × 6 MB ≈ 1.8 GB, which matches what you observed.

**2. Compressing frames in the delay queue**

You can encode each frame to JPEG bytes in memory when it goes into the queue and decode it back to a `Mat` when it comes out. JPEG at 80% quality typically compresses a 1080p frame down to 150–300 KB — a 20–40× reduction:

```csharp
// On capture — encode to JPEG bytes before queuing
byte[] Compress(Mat frame)
{
    using var buf = new VectorOfByte();
    CvInvoke.Imencode(".jpg", frame, buf,
        new KeyValuePair<ImwriteFlags, int>(ImwriteFlags.JpegQuality, 80));
    return buf.ToArray();
}

// On display — decode back to Mat
Mat Decompress(byte[] data)
{
    using var buf = new VectorOfByte(data);
    return CvInvoke.Imdecode(buf, ImreadModes.Color);
}
```

This trades CPU time (encode + decode per frame) for memory. For a 10-second buffer at 30 fps the in-memory footprint drops to roughly 50–90 MB, which is far more manageable.

If quality is less important than speed you can use PNG with `ImwriteFlags.PngCompression = 1` (fastest), or drop quality to 60–70% JPEG for an even smaller footprint.

Closing now.
